sympy/integrals/heurisch.py

```diff
@@ -137,58 +131,16 @@ def heurisch_wrapper(f, x, rewrite=False, hints=None, mappings=None, retries=3,
                    unnecessary_permutations, _try_heurisch)
     if not isinstance(res, Basic):
         return res
-    # We consider each denominator in the expression, and try to find
-    # cases where one or more symbolic denominator might be zero. The
-    # conditions for these cases are stored in the list slns.
-    slns = []
-    for d in denoms(res):
-        try:
-            slns += solve(d, dict=True, exclude=(x,))
-        except NotImplementedError:
-            pass
-    if not slns:
-        return res
-    slns = list(uniq(slns))
-    # Remove the solutions corresponding to poles in the original expression.
-    slns0 = []
-    for d in denoms(f):
-        try:
-            slns0 += solve(d, dict=True, exclude=(x,))
-        except NotImplementedError:
-            pass
-    slns = [s for s in slns if s not in slns0]
-    if not slns:
-        return res
-    if len(slns) > 1:
-        eqs = []
-        for sub_dict in slns:
-            eqs.extend([Eq(key, value) for key, value in sub_dict.items()])
-        slns = solve(eqs, dict=True, exclude=(x,)) + slns
-    # For each case listed in the list slns, we reevaluate the integral.
-    pairs = []
-    for sub_dict in slns:
-        expr = heurisch(f.subs(sub_dict), x, rewrite, hints, mappings, retries,
-                        degree_offset, unnecessary_permutations,
-                        _try_heurisch)
-        cond = And(*[Eq(key, value) for key, value in sub_dict.items()])
-        generic = Or(*[Ne(key, value) for key, value in sub_dict.items()])
+
+    def reintegrator(f, x):
+        expr = heurisch(f, x, rewrite, hints, mappings, retries, degree_offset,
+                        unnecessary_permutations, _try_heurisch)
```

Comment:
I'm not sure why any of the integration algorithms should reuse only that algorithm for the recursive integration, rather than integrate(). Does anything break for heurisch if we change this? Heurisch in particular is the slowest algorithm, which is always tried last, so it's better to avoid using it if we can. It's also unclear to me that all these arguments to heurisch should be the same for the new substituted integrand.

The only place where it might make sense to keep the same algorithm would be Risch, if we can be sure the "simplified" integral is also of the form that Risch can handle (most likely this is the case, though I'd have to look at the specific places where the divisions happen to be sure). Even then, though, Risch has many NotImplementedError holes, so it still might not make sense there.